## README

This Nextflow worfklow aligns, calls, and annotates whole genome sequencing data using Singularity. Submodules of the workflow are described and can be run independently.

Current alignment and variant calling workflow design for matched tumor-normal data generated by whole genome sequencing. 
  - 1.) QC with Fastqc
  - 2.) Trim with TrimmomaticPE
  - 3.) Fastqc on trimmed reads
  - 4.) MultiQC on all fastqc output files
  - 5.) Alignment with bwa-mem2
  - 6.) sort and index with samtools 
  - 7.) Mark duplicates with gatk MarkDuplicates
  - 8.) somatic variant calling with gatk Mutect2
  - 9.) Learn orientation bias model
  - 10.) Get pileup summaries
  - 11.) Get segmentation and contamination tables
  - 12.) Filter variant calls with gatk Mutect2 (using input files from steps 8-11)
  - 13.) Annotate variants with snpEff

Prerequisites: environment set-up [REQUEST for others' input on whether Conda installation should be included in README when workflow is publicly 
available]
  1.) Conda create -n NAME python=3.11
  2.) Install nextflow (conda install nextflow)
  2.) Install fastqc (conda install fastqc)
  3.) Install multiqc (pip install multiqc)
  4.) Install trimmomatic (conda install trimmomatic) 
  5.) Install bwa-mem2 (conda install bwa-mem2) 
  6.) Install sam-tools (conda install -c bioconda samtools)
  7.) Install gatk (conda install gatk) or download from gatk github
  8.) Install snpEff (conda install snpEff) or download from snpEff source page 
      - download reference genome with snpEff download GRCh38.86 
## Alignment workflow
[stack here]
## Variant Calling Workflow and Annotation Workflow [WIP] 
This variant calling worfklow uses Nextflow and is using Singularity. Submodules of the workflow are described below and can be run
independently assuming the necessary input files exist. A parameter file is passed as input using -params-file <my-params.json>, which can begenerated using the templating script. In the example below, parameters are passed as command line argument for to easily demonstrate usage.

## Tools
- Nextflow (DSL2)
- Singularity 
- bcftools/1.10.2
- GATK:4.4.0.0
  - GeneratePileUpSummaries
  - CalculateContamination 
  - Mutect2
  - concat
  - sort
  - index
  - normalize
  - merged
- bedtools2>=2.29.0
- SNPEff/4.5
- GNU parallel/20220522
- htslib/1.10.2

## Usage
## Run the variant calling and annotation workflow
```nextflow run WGS-variant-call-annotate-nextflow-pipeline.nf —params-file <my-params.json> -c <my-nextflow.config> -with-singularity <image.sif>```

### 1. GetPileupSummaries
*Generate read counts for estimating of contamination in the tumor tissue using GATK.*

```docker run quay:io//ohsu-comp-bio/gatk:4.4.0.0 gatk GetPileupSummaries -I tumor.bam -V gnomad.vcf -L intervals.list -O tumor_pileups.table```

```docker run quay:io//ohsu-comp-bio/gatk:4.4.0.0 gatk GetPileupSummaries -I normal.bam -V gnomad.vcf -L intervals.list -O normal_pileups.table```

### 2. Calculate Contamination
*calculate fraction of normal cell contaminants in tumor sample.*

```docker run quay.io://ohsu-comp-bio/gatk:4.4.0.0 gatk CalculateContamination -I tumor_pileups.table -matched normal_pileups.table -O contamination.table```

### 3. Split BAM files by chromosome 
_Split files by chromosome for faster processing_

```docker build -t quay.io://ohsu-comp-bio/bedtools:2.30.0 ./workdir ## WIP ```

### 4. Run Mutect2
*Call somatic variants*

```docker run quay.io://ohsu-comp-bio/gatk:4.4.0.0 gatk Mutect2 -R reference.fa -I tumor.bam -I normal.bam -normal NORMAL -pon gnomad_panel_of_normals.vcf -germline-resource exac_germline_mutation_data.vcf --f1r2-tar-gz “${file%.bam}.f1r2.tar.gz -O “${file%.bam}.unfiltered.vcf"```

### 5. Aggregate across chromosomes with bedtools 

```docker run quay.io//ohsu-comp-bio/bcftools:1.12 bcftools concat -a -f -l listSamplesToCatByChromosome -o “${file%.unf.vcf.gz}.concat.vcf" ```


### ** 6. GATK LearnReadOrientationModel **
_Learn the read orientation model to refine variant calls by removing technical artifacts._ 

```docker run quay.io://ohsu-compbio/gatk:4.4.0.0 gatk LearnReadOrientationModel -I f1r2.tar.gz -O read-orientation-model.tar.gz```

**### 7. Process VCFs with bcftools**
_Sort, index, normalize and combine (per sample) the VCF files before filtering (Step 8)_

###   7A. combine across chromosomes
```docker run quay.io//ohsu-comp-bio/bcftools:1.12 bcftools concat -a -f -l listSamplesToCatByChromosome -o “${file%.unf.vcf.gz}.concat.vcf" ```

###   7B. Sort bgzipped VCFs

``` docker run quay.io//ohsu-comp-bio/bcftools:1.12 bcftools sort “${file}.concat.vcf" -Oz -o  "${file%.unf.concat.vcf.gz}.unfiltered.sorted.vcf.gz"```

###   7C. Index files

``` docker run quay.io//ohsu-comp-bio/bcftools:1.12 bcftools index -t “${file}.unfiltered.sorted.vcf.gz" ```

### 7D. Normalize somatic valls

```docker run quay.io//ohsu-comp-bio/bcftools:1.12 normalize -i "${file}.unfiltered.sorted.vcf.gz" -o
"${file%.unfiltered.sorted.vcf.gz}.unfiltered.normalized.vcf" ``` 

**### 8. GATK FilterMutectCalls**
_Apply filters to Mutect2 variant calls_ 

```docker run quay.io://ohsu-comp-bio/gatk:4.4.0.0 gatk FilterMutectCalls -R reference.fa -V  “${file}.unfiltered.normalized.vcf" --contamination-table contamination.table --orientation-bias-artifact-priors read-orientation-model.tar.gz -O “${file%.unfiltered.normalized.vcf.gz}.filtered.vcf" ```

### **9. SNPEff Annotation**
_Annotate variants using SNPEff_ [WIP}

```docker run quay.io://ohsu-comp-bio/SNPEff:latest -v genome_version <annotated_variants.vcf> | gzip > annotated_variants.vcf.gz```



